TestBind seems flaky

Log from Travis:

```
$ make test
go test -v github.com/habitat-sh/habitat-operator/cmd/habitat-operator github.com/habitat-sh/habitat-operator/pkg/apis/habitat/v1beta1 github.com/habitat-sh/habitat-operator/pkg/client github.com/habitat-sh/habitat-operator/pkg/controller
?   	github.com/habitat-sh/habitat-operator/cmd/habitat-operator	[no test files]
=== RUN   TestRoundTrip
--- PASS: TestRoundTrip (0.06s)
	roundtrip.go:154: round tripping habitat.sh/v1beta1, Kind=Habitat
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=Habitat 
	roundtrip.go:154: round tripping habitat.sh/v1beta1, Kind=HabitatList
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
	roundtrip.go:242: 	round tripping external type habitat.sh/v1beta1, Kind=HabitatList 
PASS
ok  	github.com/habitat-sh/habitat-operator/pkg/apis/habitat/v1beta1	0.179s
?   	github.com/habitat-sh/habitat-operator/pkg/client	[no test files]
?   	github.com/habitat-sh/habitat-operator/pkg/controller	[no test files]
The command "make test" exited with 0.
$ make TESTIMAGE=habitat/habitat-operator:testing e2e
go test -v ./test/e2e/ --image "habitat/habitat-operator:testing" --kubeconfig ~/.kube/config --ip "$(minikube ip)"
=== RUN   TestBind
--- FAIL: TestBind (141.14s)
	operator_test.go:144: Configuration update did not go through. Expected: "hello from port: 6333", got: "hello from port: 4444
		"
=== RUN   TestHabitatDelete
--- PASS: TestHabitatDelete (6.02s)
=== RUN   TestPersistentStorage
--- SKIP: TestPersistentStorage (0.00s)
	operator_test.go:191: This test cannot be run successfully in our current testing setup
FAIL
exit status 1
FAIL	github.com/habitat-sh/habitat-operator/test/e2e	149.456s
make: *** [e2e] Error 1
The command "make TESTIMAGE=habitat/habitat-operator:testing e2e" exited with 2.
$ kubectl logs -lhabitat=true --tail=100
Done. Your build exited with 1.
```

--- a/test/e2e/operator_test.go
+++ b/test/e2e/operator_test.go
@@ -28,8 +28,11 @@ import (
 )
 
 const (
-	defaultWaitTime = 1 * time.Minute
-	configMapName   = "peer-watch-file"
+	serviceStartupWaitTime = 1 * time.Minute
+	secretUpdateTimeout    = 2 * time.Minute
+	secretUpdateQueryTime  = 10 * time.Second
+
+	configMapName = "peer-watch-file"
 )
 
 // TestBind tests that the operator correctly created two Habitat Services and bound them together.
@@ -97,7 +100,7 @@ func TestBind(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	time.Sleep(defaultWaitTime)
+	time.Sleep(serviceStartupWaitTime)
 
 	// Get response from Habitat Service.
 	url := fmt.Sprintf("http://%s:30001/", framework.ExternalIP)
@@ -126,22 +129,29 @@ func TestBind(t *testing.T) {
 		t.Fatalf("Could not update Secret: \"%s\"", err)
 	}
 
-	// Wait for SecretVolume to be updated.
-	time.Sleep(defaultWaitTime)
-
-	// Check that the port differs after the update.
-	body, err = utils.QueryService(url)
-	if err != nil {
-		t.Fatal(err)
-	}
-
+	ticker := time.NewTicker(secretUpdateQueryTime)
+	defer ticker.Stop()
+	timer := time.NewTimer(secretUpdateTimeout)
+	defer timer.Stop()
 	// Update the message set in the config of the habitat/bindgo-hab Go Habitat Service.
 	expectedMsg = fmt.Sprintf("hello from port: %v", 6333)
-	actualMsg = body
-	// actualMsg can contain whitespace and newlines or different formatting,
-	// the only thing we need to check is it contains the expectedMsg.
-	if !strings.Contains(actualMsg, expectedMsg) {
-		t.Fatalf("Configuration update did not go through. Expected: \"%s\", got: \"%s\"", expectedMsg, actualMsg)
+	for {
+		// Check that the port differs after the update.
+		actualMsg, err := utils.QueryService(url)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// actualMsg can contain whitespace and newlines or different formatting,
+		// the only thing we need to check is it contains the expectedMsg.
+		if strings.Contains(actualMsg, expectedMsg) {
+			break
+		}
+		select {
+		case <-timer.C:
+			t.Fatalf("Configuration update did not go through. Expected: \"%s\", got: \"%s\"", expectedMsg, actualMsg)
+		case <-ticker.C:
+		}
 	}
 }
 

--- a/test/e2e/operator_test.go
+++ b/test/e2e/operator_test.go
@@ -147,10 +147,23 @@ func TestBind(t *testing.T) {
 		if strings.Contains(actualMsg, expectedMsg) {
 			break
 		}
+
+		fail := func() {
+			t.Fatalf("Configuration update did not go through. Expected: \"%s\", got: \"%s\"", expectedMsg, actualMsg)
+		}
+
 		select {
 		case <-timer.C:
-			t.Fatalf("Configuration update did not go through. Expected: \"%s\", got: \"%s\"", expectedMsg, actualMsg)
+			fail()
 		case <-ticker.C:
+			// This is to avoid infinite loops when go
+			// decides to always pick the ticker channel,
+			// even when timer channel is ready too.
+			select {
+			case <-timer.C:
+				fail()
+			default:
+			}
 		}
 	}
 }

